### ISSUE TITLE
8088 GLaBIOS cannot boot from DTC 5150X

### ISSUE THREAD

**Issue by Teppic1** [86Box/86Box#4074]
### What happened?

The HDD isn't detected on the POST screen. It does detect the BIOS however (C800), and the drive can be used if you boot from a floppy, so the controller BIOS itself works.

This may be a GLaBIOS bug, though it is able to detect and boot the other emulated 8 bit disk controller cards I tried.

Confirmed that the DTC controller boots fine on other XT/AT machines.

![H4qQcnL](https://github.com/86Box/86Box/assets/32468801/256cf5e7-5c5a-4a4c-b0f7-be7d6df70933)

### Configuration file

```ini
[General]
vid_renderer = qt_software
force_43 = 1

[Machine]
machine = glabios
cpu_family = 8088
cpu_speed = 4772728
cpu_multi = 1
cpu_use_dynarec = 0
fpu_softfloat = 0
time_sync = local
mem_size = 640

[Video]
gfxcard = cga

[Input devices]
mouse_type = none

[Sound]
fm_driver = nuked

[Network]
net_01_link = 0
net_02_link = 0
net_03_link = 0
net_04_link = 0

[Storage controllers]
hdc = st506_xt_dtc5150x
cassette_mode = load

[Hard disks]
hdd_01_parameters = 17, 7, 987, 0, mfm
hdd_01_mfm_channel = 0
hdd_01_fn = priam170.img

[Floppy and CD-ROM drives]
fdd_01_turbo = 1
```

### Operating system

Windows 10

### CPU

i7 7500

### 86Box version

b5409

### Build architecture

Windows - x64 (64-bit)

### Build type

- [ ] New recompiler
- [ ] Debug build

### Download source

Official website (Jenkins, GitHub)

**Comment by Fenix770** [comment]
I tested and got the same issue as you!. I also I opened a similar thread about two weeks ago.

1. I think there is some BIOS limitation on the GLaBIOS. 

2. I, also, found that AMI XT works fine with the disk, but if I use the Magitronic B215 with high density floppy drives I get ROM Basic error,  if I use low density floppy drive it work. If it were a memory conflict, it would affect the low density floppy configuration as well.

Procedure
1. I booted the machine with PC-DOS 3.30, on the GLaBIOS
2. Used the DEBUG utility to do the low level format
3. I booted from MS-DOS 5.0 install disk, (also on 5.25 360kb format), MS-DOS 5.0 detects the disk size correctly, even if GLaBIOS cannot, and is bootable on AMI XT.

![Uploading Monitor_1_20240122-190252-892.png…]()

**Comment by Fenix770** [comment]
The high density controller bug is present, also on DTK PII-151B, 158B and Moster FDC. (all tested on AMI XT 8088)

1. DTK PII-151B throws the same error as the Magitronic "BASIC not present"
2. DTK PII-158B, boots fine but cannot read high density disk even with HD floppy drives
3. The Monster FDC freezes at startup

**Comment by Teppic1** [comment]
My first thought was this was likely a GLaBIOS code bug but it seemed strange to me that it has no problems booting up on other 8 bit controllers.

And it's not going to be easy to find anyone with a real XT machine with GLaBIOS installed who also has a DTC 5150X controller.

**Comment by OBattler** [comment]
Try replacing the GLaBIOS we use, which is the "emulator" version with the one designed for real hardware.

**Comment by OBattler** [comment]
OK, so this gets even more bizarre. The GLaBIOS designed for real hardware doesn't ever detect the hard disks with that controller. The "emulator" version we currently use, does, but only if a floppy is inserted into drive A:.

**Comment by Fenix770** [comment]
I also found that the DTC 5150X controller does not work properly with high density controllers. The AMI XT works with the 57mb HDD (Priam-170), but with the High density floppy controller and drive, the machine does not detect the HDD.

I, now suspect the DTC 5150X controller may have bugs, and this card was ported from PCem. PCem added it sometime around 2017

**Comment by Teppic1** [comment]
On PCem17 I tried renaming the GLaBIOS image for real hardware to the BIOS for the AMI XT (so it would load GLaBIOS instead) and the drive isn't detected there either.

**Comment by 640-KB** [comment]
**TL;DR** try these: 

https://github.com/640-KB/GLaBIOS/releases/download/v0.2.5/GLABIOS_0.2.6_8X_012324.ROM
https://github.com/640-KB/GLaBIOS/releases/download/v0.2.5/GLABIOS_0.2.6_8E_012324.ROM

I've looked at this too, and what is happening is that the call to the DTC BIOS's `INT 13h, AH=0` (reset) always returns `80h + CF` (drive timeout).  So started tracing through the DTC BIOS and noticed it seems to be using what it considered unused bytes in the BDA for its own temporary storage.  For example, during POST it appears to use the LPT port timeout bytes (`40:7A-7B`) as a flag during its setup for whether or not it is the only hard drive controller (so far). 

However, more interesting is that it also seems to use the floppy drive controller data stack buffer (`40:42-48`) as some kind of state as well.  This storage is supposed to belong to the floppy controller, but the DTC is using the last byte as part of its state machine in the hard drive reset procedure, however it also executes the floppy controller reset procedure in the middle of that.  This probably wouldn't be a problem because normally during a floppy controller reset, it wouldn't be a problem since a reset/sense only has three status bytes.  In GLaBIOS after a floppy sense and read status operation, it zeros out the remaining bytes in the FDC result stack buffer (https://github.com/640-KB/GLaBIOS/blob/f83cd3f6d0be0ac684536e486c5cfcdd3c75635c/src/GLABIOS.ASM#L7596).  The DTC BIOS then counts on this, that any floppy controller operation results are at most 3 bytes so as to not overwrite its (stolen) bytes.

Long story short...  there's probably no advantage to zero'ing out bytes from a previous operation, other than it makes it clearer for debugging.  If apparently no other BIOS does it, I'm fine with getting rid of it.

Note: even with this removed, the controller still returns a `80h` "drive not ready" during POST, so it doesn't show the drive on the POST screen. Need more research to see why that is happening as well.

**Comment by 640-KB** [comment]
> I also found that the DTC 5150X controller does not work properly with high density controllers. The AMI XT works with the 57mb HDD (Priam-170), but with the High density floppy controller and drive, the machine does not detect the HDD.
> 
> I, now suspect the DTC 5150X controller may have bugs, and this card was ported from PCem. PCem added it sometime around 2017

I see the same behavior on other XT BIOSes as well, with Multi-Floppy BIOS and DTC 5150X controller combination.

**Comment by Teppic1** [comment]
Yep, renaming the 0.2.6 BIOS above to GLABIOS_0.2.5_8E.ROM to get it to load seems to work. As you said it doesn't show on the POST screen but otherwise boots and seems to be able to read ok.

![1KEyhYl](https://github.com/86Box/86Box/assets/32468801/e1ecc35f-b59f-4e75-9475-0a02c1d748ca)
